### PR TITLE
datapath: Enable N/S LB for overlapping pod CIDR

### PR DESCRIPTION
**Pull request by aspsk**
See the commit description.

```release-note
Enable N/S LB for overlapping pod CIDR
```

### CODE REVIEW

**Review by brb — APPROVED** [review]

File: bpf/tests/tc_nodeport_lb4_nat_lb.c [hunk 1]
```diff
@@ -111,6 +111,10 @@ mock_ctx_redirect(const struct __sk_buff *ctx __maybe_unused,
 
 #define SECCTX_FROM_IPCACHE 1
 
+#ifndef SECLABEL
+#define SECLABEL 0xffff
+#endif
+
```
Comment (julianwiedmann) [inline bpf/tests/tc_nodeport_lb4_nat_lb.c]: We probably just missed to include `config_replacement.h` here (or didn't need it until now?)

File: bpf/lib/nodeport.h [hunk 2]
```diff
@@ -2706,7 +2709,7 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 						IPV4_DIRECT_ROUTING,
 						src_port,
 						tunnel_endpoint,
-						WORLD_IPV4_ID,
+						src_identity,
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h:2715]: Two blocking concerns here.

1. how do we preserve backwards compatibility for existing network policies?
2. we mustn't pass CIDR identities to other nodes, they have node-local scope. See https://github.com/cilium/cilium/pull/28441 :).

File: bpf/lib/nodeport.h [hunk 3]
```diff
@@ -2647,6 +2648,8 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 	struct iphdr *ip4;
 	__s8 ext_err = 0;
 #ifdef TUNNEL_MODE
+	__u8 cluster_id __maybe_unused = (__u8)ctx_load_meta(ctx, CB_CLUSTER_ID_EGRESS);
+	__u32 src_identity = ctx_load_meta(ctx, CB_SRC_LABEL) ?: WORLD_IPV4_ID;
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h:2652]: I think we should do this selection in `nodeport_lb4()` already. Also note that we currently use `WORLD_IPV4_ID` when [creating](https://github.com/cilium/cilium/blob/fa0037638534635471c8e45583d78f87ad78b3e1/bpf/lib/nodeport.h#L2917) the nodeport CT entry, should probably stay in sync (even though I have no clue what it's used for ...).

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
Few comments below, before I forget them ...

File: bpf/lib/nodeport.h (hunk 2 above)
Comment (brb) [reply bpf/lib/nodeport.h:2715]: Do we have a helper function to check whether sec id is a CIDR one?

Comment (julianwiedmann) [reply bpf/lib/nodeport.h:2715]: > Do we have a helper function to check whether sec id is a CIDR one?

I suspect it's this:
https://github.com/cilium/cilium/blob/0bf036effa18390c96c188be00e19cdd923050e4/bpf/lib/identity.h#L142-L143

**Comment by jibi** [comment]
superseeded by https://github.com/cilium/cilium/pull/30348

**Pull request closed without merging**